oci/archive/reader.go

```diff
@@ -0,0 +1,93 @@
+package archive
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/containers/image/v5/internal/tmpdir"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/storage/pkg/archive"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type reader struct {
+	manifest *imgspecv1.Index
+	tempDirOCIRef
+}
+
+func CreateUntarTempDirReader(src string, sys *types.SystemContext) (*reader, error) {
+	// TODO: This can take quite some time, and should ideally be cancellable using a context.Context.
+	arch, err := os.Open(src)
+	if err != nil {
+		return nil, err
+	}
+	defer arch.Close()
+
+	dst, err := ioutil.TempDir(tmpdir.TemporaryDirectoryForBigFiles(sys), "oci")
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating temp directory")
+	}
+
+	reader := reader{
+		tempDirOCIRef: tempDirOCIRef{tempDirectory: dst},
+	}
+
+	if err := archive.NewDefaultArchiver().Untar(arch, dst, &archive.TarOptions{NoLchown: true}); err != nil {
+		if err := reader.tempDirOCIRef.deleteTempDir(); err != nil {
+			return nil, errors.Wrapf(err, "error deleting temp directory %q", dst)
+		}
+		return nil, errors.Wrapf(err, "error untarring file %q", dst)
+	}
+
+	indexJSON, err := os.Open(filepath.Join(dst, "index.json"))
+	if err != nil {
+		return nil, err
+	}
+	defer indexJSON.Close()
+	reader.manifest = &imgspecv1.Index{}
+	if err := json.NewDecoder(indexJSON).Decode(reader.manifest); err != nil {
+		return nil, err
+	}
+
+	return &reader, nil
+}
+
+// ReferenceWrapper includes the reference for the image and image dest name
+type ReferenceWrapper struct {
+	DestName       string
+	ImageReference types.ImageReference
+}
+
+// List returns a list of ReferenceWrapper for images in the reader.
+// the ImageReferences of the ReferenceWrapper are valid only until the Reader is closed.
+func (r *reader) List() ([]ReferenceWrapper, error) {
```

Comment:
Looking at `docker-archive`, I think we should not create a custom `ReferenceWrapper` but return a `[][]types.ImageReference`. With the first array determining the amount of images and the second one the "names".

`docker-archive` supports loading by index and by name and I think that's what we should do here as well to be consistent. The docker-archive parts are in `docker/archive/reader.go`.

Reply:
@vrothberg Could you elaborate on what's loading by index and name?

Reply:
A `poadman load -i` always expects a file, so we can't use the transport syntax. But we can individually pull/extract images from an archive. To extract an image (note we can do this with `podman pull` as well), we need to specify which one. We can specify the image either via a docker-reference (without digest) or by it's index in the list of manifests.

Examples are the man page (https://github.com/containers/image/blob/master/docs/containers-transports.5.md#docker-archivepathdocker-referencesource-index) or the tests in podman that exercise this code (https://github.com/containers/podman/commit/7fea46752cbf#diff-36d4ee0b4d8927b38ddbff8a5648df51ebf0e9289fcb0d37e0dedc2ffa3ee161R173).